src/connection/query.ts

```diff
@@ -12,28 +12,57 @@ export class Query<T> {
   private correlationId: string;
   private sql: string;
   private isPrepared: boolean = false;
-  private parameters: any[];
+  private parameters: any[]|undefined;
   private rowsToFetch: number = 100;
   private isCLCommand: boolean;
   private state: QueryState = QueryState.NOT_YET_RUN;
 
-  constructor(private job: SQLJob, query: string, opts: QueryOptions = {isClCommand: false, parameters: undefined}) {
+  public autoClose: boolean | undefined;
+
+  constructor(private job: SQLJob, query: string, opts: QueryOptions = { isClCommand: false, parameters: undefined, autoClose: false }) {
     this.job = job;
     this.isPrepared = (undefined !== opts.parameters);
     this.parameters = opts.parameters;
     this.sql = query;
     this.isCLCommand = opts.isClCommand;
+    this.autoClose = opts.autoClose;
+
     Query.globalQueryList.push(this);
   }
+
   public static byId(id: string) {
     return (undefined === id || '' === id) ? undefined : Query.globalQueryList.find(query => query.correlationId === id);
   }
+
+  public static getOpenIds(forJob?: string) {
+    return this.globalQueryList
+      .filter(q => q.job.id === forJob || forJob === undefined)
+      .map(q => q.correlationId);
+  }
+
+  public static async cleanup() {
+    let closePromises = [];
+
+    // First, let's check to see if we should also cleanup
+    // any cursors that remain open, and we've been told to close
+    for (const query of this.globalQueryList) {
+      if (query.autoClose) {
+        closePromises.push(query.close())
+      }
+    };
+
+    await Promise.all(closePromises);
+
+    // Automatically remove any queries done and dusted. They're useless.
+    this.globalQueryList = this.globalQueryList.filter(q => q.getState() !== QueryState.RUN_DONE);

```

Comment:
Is there any flows here the UI might still be open for one of these queries?

Reply:
Not that I am aware off. Really if the value is `RUN_DONE` then we should be removing it from the array. `cleanup` will set some queries to `RUN_DONE` also.

Reply:
I agree (and this, I think, is tied to the future multi-tabs discussion)